Crash on app startup - attempting to retrieve servers from empty database

Looks like wiping the database on upgrade (discussed in https://github.com/CUTR-at-USF/OpenTripPlanner-for-Android/pull/306#issuecomment-51106722) isn't currently a clean solution.  From a quick look at the code, it seems that since there is a preference `OTPApp.PREFERENCE_KEY_SELECTED_SERVER` that is set after the server is selected, if this is set to a non-zero server ID then the app attempts to grab that ID from the database, resulting in a NPE:

```
java.lang.RuntimeException: Unable to start activity ComponentInfo{edu.usf.cutr.opentripplanner.android/edu.usf.cutr.opentripplanner.android.MyActivity}: java.lang.NullPointerException
            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2216)
            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2265)
            at android.app.ActivityThread.access$800(ActivityThread.java:144)
            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1205)
            at android.os.Handler.dispatchMessage(Handler.java:102)
            at android.os.Looper.loop(Looper.java:136)
            at android.app.ActivityThread.main(ActivityThread.java:5140)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:515)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:795)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:611)
            at dalvik.system.NativeStart.main(Native Method)
     Caused by: java.lang.NullPointerException
            at edu.usf.cutr.opentripplanner.android.model.Server.<init>(Server.java:85)
            at edu.usf.cutr.opentripplanner.android.fragments.MainFragment.updateSelectedServer(MainFragment.java:2494)
            at edu.usf.cutr.opentripplanner.android.fragments.MainFragment.onActivityCreated(MainFragment.java:557)
            at android.support.v4.app.Fragment.performActivityCreated(Fragment.java:1508)
            at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:958)
            at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1115)
            at android.support.v4.app.BackStackRecord.run(BackStackRecord.java:682)
            at android.support.v4.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1478)
            at android.support.v4.app.FragmentActivity.onStart(FragmentActivity.java:570)
            at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1171)
            at android.app.Activity.performStart(Activity.java:5241)
            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2179)
            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2265)
            at android.app.ActivityThread.access$800(ActivityThread.java:144)
            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1205)
            at android.os.Handler.dispatchMessage(Handler.java:102)
            at android.os.Looper.loop(Looper.java:136)
            at android.app.ActivityThread.main(ActivityThread.java:5140)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:515)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:795)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:611)
            at dalvik.system.NativeStart.main(Native Method)
```

@vreixo Do you think the best solution for this is to reset this preference to `0` when we detect that the app `version_code` is incremented?

--- a/opentripplanner-android/src/main/AndroidManifest.xml
+++ b/opentripplanner-android/src/main/AndroidManifest.xml
@@ -16,10 +16,10 @@
 -->
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
           package="edu.usf.cutr.opentripplanner.android"
-          android:versionCode="12"
-          android:versionName="2.0.5">
+    android:versionCode="13"
+    android:versionName="2.1.0">
 
-    <uses-sdk
+<uses-sdk
             android:minSdkVersion="8"
             android:targetSdkVersion="13"/>
 

--- a/opentripplanner-android/src/main/java/edu/usf/cutr/opentripplanner/android/OTPApp.java
+++ b/opentripplanner-android/src/main/java/edu/usf/cutr/opentripplanner/android/OTPApp.java
@@ -226,6 +226,8 @@ public class OTPApp extends Application {
 
     public static final String PREFERENCE_KEY_APP_VERSION = "app_version";
 
+    public static final String PREFERENCE_KEY_EXECUTED_VERSION_CODE_13 = "executed_version_13";
+
     private static Server selectedServer;
 
     public static final String TAG = "OTP";

--- a/opentripplanner-android/src/main/java/edu/usf/cutr/opentripplanner/android/fragments/MainFragment.java
+++ b/opentripplanner-android/src/main/java/edu/usf/cutr/opentripplanner/android/fragments/MainFragment.java
@@ -2519,7 +2519,7 @@ public class MainFragment extends Fragment implements
 
             Log.v(OTPApp.TAG, "Now using OTP server: " + server.getRegion());
         } else {
-            Log.v(OTPApp.TAG, "Server not selected yet, should be first start");
+            Log.v(OTPApp.TAG, "Server not selected yet, should be first start or app update");
             return;
         }
         if (server.getOffersBikeRental()){
@@ -4013,11 +4013,20 @@ public class MainFragment extends Fragment implements
             editor.commit();
             mNewAppVersion = newVersionCode != oldVersionCode;
 
-            if (mNewAppVersion) {
+            /**
+             * Special handling for introduction of PREFERENCE_KEY_APP_VERSION - see #309
+             * Otherwise, mNewVersion will be false the first time we execute version_code 13
+             * when installed as an update to version_code 12.
+             */
+            boolean executedVersion13 = mPrefs.getBoolean(OTPApp.PREFERENCE_KEY_EXECUTED_VERSION_CODE_13, false);
+
+            if (mNewAppVersion || !executedVersion13) {
                 Log.d(OTPApp.TAG, "Updating from app version " + oldVersionCode + " to " +
                         newVersionCode);
                 // Erase selected server, so server selection is run after an app update (#309)
                 editor.putLong(OTPApp.PREFERENCE_KEY_SELECTED_SERVER, 0);
+                // We've executed version_code 13 or higher once, so set the preference
+                editor.putBoolean(OTPApp.PREFERENCE_KEY_EXECUTED_VERSION_CODE_13, true);
                 editor.commit();
             }
         } catch (PackageManager.NameNotFoundException e) {

